packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts

```diff
@@ -75,16 +101,172 @@ export class CheckpointService implements TreeDataProvider<BaseNode> {
     return element.getChildren();
   }
 
-  public addCheckpointNode(
-    sourceFile: string,
-    typeRef: string,
-    line: number
-  ): CheckpointNode {
-    const cpNode = new CheckpointNode(sourceFile, typeRef, line);
+  // There's a limit to the number of checkpoints for a given user and
+  // that number is 5.
+  public canAddCheckpointNote(): Boolean {
+    return this.getChildren().length < MAX_ALLOWED_CHECKPOINTS;
+  }
+
+  public async addCheckpointNode(
+    uriInput: string,
+    sourceFileInput: string,
+    typeRefInput: string,
+    lineInput: number,
+    makeCommandCall = true
+  ): Promise<CheckpointNode> {
+    // Before creating a new checkpoint node, check and see if one with the
+    // URI and line number already exists. If one already exists then return
+    // that one, otherwise create one, add it to the list and then return it
+    const cpTemp = this.returnCheckpointNodeIfAlreadyExists(
+      uriInput,
+      lineInput
+    );
+    if (cpTemp) {
+      return cpTemp;
+    }
+    const cpNode = new CheckpointNode(
+      uriInput,
+      sourceFileInput,
+      typeRefInput,
+      lineInput
+    );
+    // This is here for testing purposes. The command is being tested separately
+    if (makeCommandCall === true) {
+      await this.executeCreateApexExecutionOverlayActionCommand(cpNode);
+    }
     this.checkpoints.push(cpNode);
     this._onDidChangeTreeData.fire();
     return cpNode;
   }
+
+  // For checkpoints, there should only ever be one checkpoint for a given uri/line
+  // combination. ApexExecutionOverlayAction will not allow multiple actions for the
+  // same type/line.
+  private returnCheckpointNodeIfAlreadyExists(
+    uriInput: string,
+    lineInput: number
+  ): CheckpointNode | undefined {
+    for (const cp of this.checkpoints) {
+      if (cp.getCheckpointLineNumber() === lineInput) {
+        if (
+          cp.getCheckpointUri().toLocaleLowerCase() ===
+          uriInput.toLocaleLowerCase()
+        ) {
+          return cp;
+        }
+      }
+    }
+    // If we didn't find anything then just return undefined
+    return undefined;
+  }
+
+  public async deleteCheckpointNode(
+    uriInput: string,
+    lineInput: number,
+    makeCommandCall = true
+  ): Promise<boolean> {
```

Comment:
The boolean value doesn't seem to be used anywhere so it might be better to make this return a void promise.

Reply:
Done.